### ISSUE TITLE
Nearest neighbor in python

### ISSUE THREAD

**Issue by maxnoe** [matloff/R-vs.-Python-for-Data-Science#9]
> For example, I once needed code to do fast calculation of nearest-neighbors of a given data point

First google hit for "nearest neighbor" is scikit-learns k nearest neighbor classifier, which links to the `BallTree` and `KDTree` classes it uses for neighbor finding:

* https://scikit-learn.org/stable/modules/neighbors.html
* https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.BallTree.html#sklearn.neighbors.BallTree
* https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KDTree.html#sklearn.neighbors.KDTree

Little more googling about those algorithms gives you:
* https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.cKDTree.html

Also, researching nearest neighbors in generall, will give you the answer, that a KDTree is one of the most efficient algorithms to do this and googling 
`kdtree python` will yield the scipy implementation as first hit.

**Comment by matloff** [comment]
I did make a point of saying I wasn't interested in the classification app. Thanks for the tip on kdtree. Doesn't surprise me but it didn't come up in my PyPI search, which was my pt.

**Comment by maxnoe** [comment]
> I did make a point of saying I wasn't interested in the classification app. 

Yes, but the classification app *uses* and directly links to what you were looking  for.
